### PR TITLE
Hide launcher complexity from plugin.

### PR DESCRIPTION
**Pull request by xyan0607**

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
I was not watching this repo, hence I've missed it. Sorry

File: core/src/main/java/jenkins/model/logging/LoggingDefinitionLauncherWrapper.java [hunk 1]
```diff
@@ -23,10 +23,160 @@
...
+            MasterToSlaveCallable<RemoteProcess, IOException> {
+
+            private final List<String> cmd;
+            private final boolean[] masks;
+            private final String[] env;
+            private final InputStream in;
+            private final OutputStream out;
+            private final OutputStream err;
+            private final String workDir;
+            private final TaskListener listener;
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/logging/LoggingDefinitionLauncherWrapper.java]: I'm not sure we want to propagate it. I'll check it

File: core/src/main/java/jenkins/model/logging/LoggingDefinitionLauncherWrapper.java [hunk 2]
```diff
@@ -23,10 +23,160 @@
...
+            private final boolean reverseStdin, reverseStdout, reverseStderr;
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/logging/LoggingDefinitionLauncherWrapper.java]: Effectively we do not use these flags (See TODO above).
Maybe we want to just drop them

File: core/src/main/java/jenkins/model/logging/LoggingMethod.java [hunk 3]
```diff
@@ -16,32 +20,52 @@
...
-          @Nonnull Run<?,?> run, @Nonnull Node node); 
-  
-  /** 
-   * Fallback Logging methods for jobs, which do not define the implementation. 
-   */
-  public static final LoggingMethod NOOP = new NoopLoggingMethod();
-  private static class NoopLoggingMethod extends LoggingMethod {
+
+    /**
+     * Decorates logging on the Jenkins master side.
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/logging/LoggingMethod.java]: Seems I messed up the formatting in PoC

File: core/src/main/java/jenkins/model/logging/LoggingMethod.java [hunk 4]
```diff
@@ -16,32 +20,52 @@
...
+            return new DefaultLocalLauncher(original);
+        } else {
+            return new DefaultRemoteLauncher(original, run, this);
+        }
+    }
+
+    /**
+     * Provides the output stream for given run.
+     * @param run Run
+     * @return
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/logging/LoggingMethod.java]: Fine for prototype, but does not work for production code. Javadoc linter returns errors by default if tags are not documented like <code>returns</code> entries

**Comment by oleg-nenashev** [comment]
IMHO it can be merged for prototyping purposes

File: core/src/main/java/jenkins/model/logging/LoggingMethod.java (hunk 3 above)
Comment (xyan0607) [inline core/src/main/java/jenkins/model/logging/LoggingMethod.java]: Not a big deal. I used 4 spaces for a tab. Should I use 2 spaces?

File: core/src/main/java/jenkins/model/logging/LoggingMethod.java (hunk 4 above)
Comment (xyan0607) [inline core/src/main/java/jenkins/model/logging/LoggingMethod.java]: Done

File: core/src/main/java/jenkins/model/logging/LoggingDefinitionLauncherWrapper.java (hunk 2 above)
Comment (xyan0607) [inline core/src/main/java/jenkins/model/logging/LoggingDefinitionLauncherWrapper.java]: Sure, I just removed them

**Comment by xyan0607** [comment]
I think it's good to hide the launcher complexity from the extension, otherwise, LoggingMethod will be a lot like BuildWrapper since BuildWrapper has the two methods LoggingMethod has.

**Pull request merged**